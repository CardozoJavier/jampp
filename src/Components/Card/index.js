export { default as Card } from './Card';
export { default as ExpandableCard } from './ExpandableCard';
