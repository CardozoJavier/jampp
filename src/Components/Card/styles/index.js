export { CardTitle } from './CardTitle.styles';
export { CardContainer } from './CardContainer.styles';
export { CardDescription } from './CardDescription.styles';
export { CardHeader } from './CardHeader.styles';
