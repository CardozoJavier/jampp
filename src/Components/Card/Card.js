import React from 'react';
import PropTypes from 'prop-types';
import { CardContainer, CardTitle, CardDescription, CardHeader } from './styles';

/**
 * Card component can be called with
 * @param {Node} children - (Optional)  It's the content to be displayed into Card.
 * @param {String} title - (Optional) It's title to be displayed in card.
 * @param {String} description - (Optional) It's description to be displayed in card below of title.
 * @param {String} width - (Optional) It's the width of the card container.
 * @param {String} padding - (Optional) It's the padding of the card container.
 * @return {React Component} A view for card with title and description.
 */
const Card = ({ children, title, description, width, padding }) => (
  <CardContainer width={width} padding={padding}>
    {title &&
      <CardHeader flexDirection={description ? 'column' : 'row'}>
        <CardTitle>{ title }</CardTitle>
        {description && <CardDescription>{ description }</CardDescription>}
      </CardHeader>
    }
    {children}
  </CardContainer>
);

Card.propTypes = {
  children: PropTypes.node,
  title: PropTypes.string,
  description: PropTypes.string,
};

Card.defaultProps = {
  children: null,
  title: '',
  description: '',
};

export default Card;
