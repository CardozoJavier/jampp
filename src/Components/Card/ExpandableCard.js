import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { ExpandableCardContainer, CardTitle, ExpandableCardDescription, ExpandableCardHeader, ChildrenContainer } from './styles';
import { DownChevronIcon, BoldAddIcon, TrashIcon } from '../UI/Icons';
import { getClassName } from '../../utils';
import { palette } from '../styles';
import { Button } from '../Button';
import { ModalContainerB, ModalRowContainer } from '../Modal/styles';
import { Modal } from '../Modal';
const { gray, black } = palette;

const expandableCardClassesName = {
  opennedClassName: 'expandable card--default__expand',
  closedClassName: 'expandable card--default__collapse'
};

/**
 * ExpandableCard component can be called with
 * @param {Node} children - (Optional)  It's the content to be displayed into Card.
 * @param {String} title - (Optional) It's title to be displayed in card.
 * @param {String} description - (Optional) It's description to be displayed in card below of title.
 * @param {String} width - (Optional) It's the width of the card container.
 * @param {String} padding - (Optional) It's the padding of the card container.
 * @return {React Component} A view for card with title and description.
 */
const ExpandableCard = ({ children, title, description, width, padding, borderBottom, addStructure }) => {
  const childrenCloned = React.cloneElement(children);
  const { opennedClassName, closedClassName } = expandableCardClassesName;

  const [className, setClassName] = useState(closedClassName);  
  const [expand, setExpand] = useState(false);
  const [structure, setStructure] = useState([children]);
  const [modal, setModal] = useState(false);
  
  const toggleToClassName = getClassName(className, opennedClassName, closedClassName);

  const handleClick = () => {
    setClassName(toggleToClassName);
    setExpand(!expand);
  };

  /**
   * Show modal to confirm remove structure.
   */
  const displayModal = (key) => {
    setModal(renderModal(key));
  }

  /**
   * Modify 'id' prop from origin children element.
   */
  const settingUniqueStructureId = (clone) => (
    clone.props.children.map(child => (
      React.cloneElement(child, {
        id: Math.random().toString(),
        key: Math.random().toString(),
      })
    ))
  );

  /**
   * Add duplicated children structure to be rendered.
   */ 
  const handleAddStructure = () => {
    const key = Math.random().toString();
    const structuresArray = Array.isArray(structure) ? [...structure] : [structure];
    const newStructure = React.cloneElement(childrenCloned, {
      key,
      id: key,
      className: 'trash-icon__visible',
      children: settingUniqueStructureId(childrenCloned),
    });
    
    // Add Trash icon in duplicated structures to allowing remove them.
    newStructure.props.children.unshift(
      <TrashIcon props={{
          width: '18px',
          height: '18px',
          fill: gray.g3,
          margin: '0 0 0 auto',
          cursor: 'pointer',
          onClick: () => displayModal(key),
          hover: black,
        }}
      />
    );

    structuresArray.push(newStructure);
    setStructure(structuresArray);
  };

  /**
   * Remove from DOM the structure selected.
   */
  const handleRemove = (confirm, key) => {
    if (confirm === 'remove-structure') {
      const structureDeleted = structure.filter(struct => key !== struct.key);
      setStructure(structureDeleted);
      setModal(null);
    } else {
      setModal(null);
    }
  };

  /**
   * Render modal to confirm remove structure.
   */
  const renderModal = (key) => (
    <ModalContainerB width="480px" position="absolute" top="50%" left="25%">
      <Modal title="Lorem ipsum dolor sit amet, consectetur adipiscing ?" icon={TrashIcon}>
        <ModalRowContainer>
          <Button label="No, cancel" type="secondary-gray-medium" onClick={() => handleRemove(null, key)} />
          <Button label="Yes, delete" type="secondary-red-medium" onClick={() => handleRemove('remove-structure', key)} />
        </ModalRowContainer>
      </Modal>
    </ModalContainerB>
  );

  return (
    <ExpandableCardContainer width={width} padding={padding} className={className}>
      {title &&
        <ExpandableCardHeader flexDirection={description ? 'column' : 'row'} borderBottom={borderBottom}>
          <CardTitle>{ title }</CardTitle>
          {description && <ExpandableCardDescription>{ description }</ExpandableCardDescription>}
          <DownChevronIcon props={{
            width: '24px',
            height: '24px',
            className,
            onClick: handleClick,
            cursor: 'pointer',
            fill: gray.g4,
            }}
          />
        </ExpandableCardHeader>
      }
      <ChildrenContainer className={className}>
        {structure}
        {modal}
        {addStructure &&
          <Button onClick={handleAddStructure} label="Add second tracking" type="duplicate-structure" icon={BoldAddIcon} />
        }
      </ChildrenContainer>
    </ExpandableCardContainer>
  );
};

ExpandableCard.propTypes = {
  children: PropTypes.node,
  title: PropTypes.string,
  description: PropTypes.string,
  width: PropTypes.string,
  padding: PropTypes.string,
  borderBottom: PropTypes.string,
};

ExpandableCard.defaultProps = {
  children: null,
  title: '',
  description: '',
  width: '',
  padding: '',
  borderBottom: '',
};
/**
 * TODO:
 *  Resolver problema al eliminar las estructuras: con cada llamado a "handleRemove" se tiene la referencia al state desactualizada.
 */
export default ExpandableCard;
