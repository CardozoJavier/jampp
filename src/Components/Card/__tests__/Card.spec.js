/* eslint-disable no-undef */
import React from 'react';
import { render, screen } from '@testing-library/react';
import Card from '../Card';

describe('<Card/> component', () => {
  const title = 'Test title';
  const description = 'Test long description';
  const paragraph = 'Paragraph element like children prop';
  const children = <p>{paragraph}</p>;

  test('should render title', () => {
    render(<Card title={title} />);
    expect(screen.getByText(title)).not.toBeNull();
  });

  test('should render description', () => {
    render(<Card title={title} description={description} />);
    expect(screen.getByText(description)).not.toBeNull();
  });

  test('not should render title neither description', () => {
    render(<Card />);
    expect(screen.queryByText(title)).toBeNull();
    expect(screen.queryByText(description)).toBeNull();
  });

  test('should render children', () => {
    render(<Card>{children}</Card>);
    expect(screen.getByText(paragraph)).not.toBeNull();
  });

  test('should match snapshot', () => {
    const { container } = render(<Card title={title} description={description}>{children}</Card>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
