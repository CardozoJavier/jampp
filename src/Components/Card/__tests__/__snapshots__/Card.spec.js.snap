// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> component should match snapshot 1`] = `
<div
  class="sc-bwzfXH ckYFdM"
  width=""
>
  <div
    class="sc-bxivhb cHFtKl"
  >
    <p
      class="sc-bdVaJa jyIbrl"
    >
      Test title
    </p>
    <p
      class="sc-htpNat eqHuob"
    >
      Test long description
    </p>
  </div>
  <p>
    Paragraph element like children prop
  </p>
</div>
`;
